RoaringBitmap/src/main/java/org/roaringbitmap/longlong/Roaring64Bitmap.java

```diff
@@ -863,10 +864,12 @@ public void advanceIfNeeded(long minval) {
         }
       }
```

Comment:
the only thing I honestly think I notice / question is on line 861 ( which I am not able to select ) the 
```
            } else {         
     return;      
      }
```
this is the else to the keyIte.hasNext() being false, I question if the "high level interator" after the seek has no next, is `charIterator` is not set null.

thus a call to hasNext(), which will check the old stale `charIterator` might say yes.

unless

Reply:
I agree. I'm setting `charIterator = null;` there, so that `hasNext()` will fail there. I'm also thinking we actually gotta set `hasNextCalled = false` there, to avoid a potential `NullPointerException` in `next()` and `peekNext()` afterwards.

Reply:
when I rewrote it today, I stripped all the `hasNextCalled` variables, and always just called `hasNext()` at the correct time to avoid stale variables. To which end I am not sure if `hasNextCalled` is there as a "performance improvement" but it doesn't seem like it makes the code stable verse the "cost" of calling `charIterator != null && charIterator.hasNext()`.

Which speaking of if hasNext is reordered to have the return true path first, then the other to larger iterate across the outer iterators can be just one while block, as they are the same body.

Reply:
You are right. Reordering and dropping `hasNextCalled` makes the code a lot simpler at the cost of a few extra comparisons. I feel this is worth it, considering how complicated the rest is already. What do you think @weijietong ?